### PR TITLE
Migrate add_to_log() to Event classes

### PR DESCRIPTION
**Pull request by nadavkav**
Hi Andrew,

While trying to revive your project, I needed to update the module to support Events
So here is a patch to update module's events. Hope I did not miss anything :-)

Please review and consider updating your repo.

Nadav

### CODE REVIEW

File: version.php [hunk 1]
```diff
@@ -29,7 +29,7 @@
 
 defined('MOODLE_INTERNAL') || die();
 
-$module->version   = 2014101300;      // The current module version (Date: YYYYMMDDXX)
+$module->version   = 2014101302;      // The current module version (Date: YYYYMMDDXX)
```
Comment (garemoko) [inline version.php]: Please update this to 2015030200 or similar to reflect the later date this was updated on.

File: view.php [hunk 2]
```diff
@@ -44,9 +44,17 @@
 }
 
 require_login($course, true, $cm);
-$context = get_context_instance(CONTEXT_MODULE, $cm->id);
-
-add_to_log($course->id, 'tincanlaunch', 'view', "view.php?id={$cm->id}", $tincanlaunch->name, $cm->id);
+$context = context_module::instance($cm->id);
+
+// Trigger module viewed event.
+$event = \mod_scorm\event\course_module_viewed::create(array(
```
Comment (garemoko) [inline view.php]: This does not look right - should it be mod_tincanlaunch ?

**Comment by garemoko** [comment]
Thanks so much for this! I made some comments on the code. Once these are resolved I'll make some time to test. 

If anybody else wants to help out by installing and test this after the changes have been made (@davidpesce ?), the code is here: https://github.com/nadavkav/moodle-mod_tincanlaunch/tree/nadavkav_events

@nadavkav what does this actually fix and in what versions of Moodle? Does it break any older versions?

File: completion_test.php [hunk 3]
```diff
@@ -76,7 +76,7 @@ function tincanlaunch_get_completion_state_test($course,$cm,$userid,$type) {
     	$areAnyStatementsReturned = tincanlaunch_check_statements($tincanlaunchsettings['tincanlaunchlrsendpoint'], $tincanlaunchsettings['tincanlaunchlrslogin'], $tincanlaunchsettings['tincanlaunchlrspass'], $tincanlaunchsettings['tincanlaunchlrsversion'], $tincanlaunch->tincanactivityid, tincanlaunch_getactor(), $tincanlaunch->tincanverbid);
 		
 		echo ('<p>LRS response:</p><pre>');
-		print_r($areAnyStatementsReturned y);
+		print_r($areAnyStatementsReturned);
```
Comment (garemoko) [inline completion_test.php]: This whole block (4 lines) looks like debugging code that's ben mistakenly left in. It can be deleted.

**Comment by scara** [comment]
@garemoko: https://docs.moodle.org/dev/Event_2 is available since 2.5 but stabilized in 2.6 while logging, https://docs.moodle.org/dev/Logging_2, has been introduced in 2.7.

`add_to_log()` has been deprecated since 2.7 and you could still use it only when the administrator activates the so called _legacy logging_. More details in: https://docs.moodle.org/dev/Migrating_logging_calls_in_plugins.

The best approach IMHO is to create a `legacy_logging` branch with the current code, with no restrictions on the required Moodle version but a note in the README about the need to activate the legacy logging for 2.7 onwards and let Nadav's code land into `master`.
Otherwise, you could create a `MOODLE_2n_STABLE` branch, for each supported version.

File: version.php (hunk 1 above)
Comment (nadavkav) [inline version.php]: Fixed

File: view.php (hunk 2 above)
Comment (nadavkav) [inline view.php]: Fixed

**Comment by nadavkav** [comment]
I am using latest master with this version of the module, so I was not really looking back on old versions.
Indeed, @scara 's suggestions are the right way to go forward and to support old (before 2.7) Moodle versions too.

btw, I am moving on to update the grades code and then get some inspiration from the block tincanlearnerstream code and make a grading page.

**Comment by scara** [comment]
@nadavkav: `$module->requires = 2013051403;` should be update to `$module->requires = 2014051200;` in order to set _2.7_ as the minimum required version (Refs.: https://docs.moodle.org/dev/Releases#Moodle_2.7).

**Comment by davidpesce** [comment]
Yeah, I can help out with testing this. I've got a bunch of Moodle versions running on different servers.

Any specific settings you're looking for testing (ie legacy logging)? Or just functionality with an LRS?

**Comment by nadavkav** [comment]
You are right @scara . Fixed.

**Comment by garemoko** [comment]
Thanks @scara - I have created a MOODLE_25_STABLE branch. 'Supported' may be a strong word, but I'll refer Moodle 2.4-2.6 users to that branch. 

@nadavkav Thanks for all your additional work on this. Please can I also ask that you update readme.md to reflect the version change. Did you also see my comment above about deleting the debugging code that you fixed the typo in? If that's ok. 

@davidpesce Sounds like this just needs testing on Moodle 2.7+. We need to test that the new code doesn't break the plugin and that the logging and events work as they should. @scara @nadavkav - any ideas how to test events/logging? There's a new events/logging interface in 2.8 I think that you might be able to make use of for that version.

**Comment by scara** [comment]
What about https://docs.moodle.org/dev/Migrating_logging_calls_in_plugins#Validation_and_testing ?

**Comment by nadavkav** [comment]
@garemoko , I have updated the readme.md file with a new section "changelog", Hope is what you ment. Usually, there is a separate file for this called Changelog. but the readme seems like a good place for this too.

Following @scara advice, I read through the documentation about testing the events which lead me to remove the redundant validate_data() function and cleanup the event triggering code to the minimal parameters necessary. I have manually tested and verified that the events show up in the Course Administration > Reports > Logs and the Site Administration > Reports > Events list, but did not add the automatic unit test file events_test.php , yet. ( I will do it later, I am currently focusing on the grading code...)

**Comment by davidpesce** [comment]
OK, I've tested the latest on nadavkav's latest nadavkav_events branch. Running Moodle 2.8.3+ (Build: 20150225)

Plugin works as expected on both SCORMcloud and LearningLocker.

In Course Admin > Reports > Logs, I'm showing three types of events:
- Course module viewed
- Activity launched
- Activity completed

Within Site Admin > Reports > Events, I see four events:
- Activity completed
- Activity launched
- Course module instance list viewed
- Course module viewed

**Comment by davidpesce** [comment]
One thing that was odd, was that every time next is clicked in the content, it logs an activity complete event. It is using the Moodle content ID for the entire course, so it looks like:

```
The user with id '2' completed the activity with id '2' for the tincan with course module id '14'.
The user with id '2' completed the activity with id '2' for the tincan with course module id '14'.
The user with id '2' completed the activity with id '2' for the tincan with course module id '14'.
```

Is there any way to use the TC ID to show what is being viewed? Is that even desired?

**Comment by scara** [comment]
@davidpesce: I think you're right but IMHO it is out of the scope of this issue and it could be better to address it under a new issue, about what data should belong to the logging of each event.

**Comment by davidpesce** [comment]
Sounds fair to me.

**Comment by nadavkav** [comment]
I can change the reporting data very easy, please advise what variable do you expect to see in the report.

Btw, does anyone of you have a Storyline 2 tincan published package that reports grades which I can use for testings?

**Comment by nadavkav** [comment]
And... @davidpesce , I was testing it using WAX LRS with the same results.

**Comment by garemoko** [comment]
Is completion tracking now tided to logging in Moodle 2.7+ ?

If so, I think the cause of the issue reported by @davidpesce is that the completion logging is in the wrong place in https://github.com/nadavkav/moodle-mod_tincanlaunch/blob/nadavkav_events/completion_check.php It should be within the [if at the bottom of the page](https://github.com/nadavkav/moodle-mod_tincanlaunch/blob/nadavkav_events/completion_check.php#L60). 

To be fair, I guess this is an existing bug but it's more pronounced with these changes. 

Thanks for adding the changelog to the readme. There's also a specific mention of Moodle 2.5 higher up - would be good to remove that if you have time. 

Thanks!

**Comment by davidpesce** [comment]
@nadavkav Yeah, I've got a Storyline 2 file you can use (link below). Let me know when you've got it and I'll remove it:

(Deleted)

**Comment by nadavkav** [comment]
Thanks @davidpesce , the SL 2 file works (the flash version, actually. the story_html5.html does not send xAPI statements to the LRS, not sure why)

Here are the statements from the WAX LRS:
image

EDIT: @garemoko , You a right. I missed the "launched" verb. it's there :-)

**Comment by garemoko** [comment]
The Moodle should be sending "launched" not "attempted". Guess that's lower down than the screenshot :-)

**Comment by nadavkav** [comment]
Regarding Moodle's internal "completion"...
It's a little problematic to calculate/set it since the "Course package" is launched outside of Moodle and is disconnected from it. Even if it sends a "completed" verb statement to the LRS... Moodle does not know about it. Guessing the user finished the activity just because the user is back to Moodle or the activity's window is closed is not an option for us, On the long run.

We can have the "cron" function periodically query the LRS for a completion verb of the specific user in that specific activity. What do you think?

Currently, while developing the grades page for the activity, I have added a "sync" button that enables the teacher to send a completion check query to the LRS and get the latest statues on all users.
And another "full details" button on each user row to get the detailed list of statements for that specific user (as seen in the previous comment)

This is just to show feasibility, but I think we should discuss it more, especially, after I get some insights from teachers that intend to use it. What do you think?

**Comment by nadavkav** [comment]
@garemoko , You a right. I missed the launched verb. it's there :-)

**Comment by garemoko** [comment]
There is already a couple of mechanisms that calls completion_check.php at different points to check for completion - that's actually the purpose of that file. You've notice there's a javascript loop that periodically calls that page whilst the learner is in the activity and that it gets called when the learner enters or leaves the activity. 

It's not fool proof though and the sync button will help to catch cases that fall through the gaps. 

The ideal scenario would be to have a Moodle endpoint that can receive forwarded statements from an LRS (not a full LRS in Moodle, just an endpoint that catches the POST) and checks for completion based on that.

**Comment by nadavkav** [comment]
Ok. now I understand why I am getting a lot of "activity completed" events in the course's log report.
I hope I fix it by moving the event trigger inside the "completed" IF statement.

**Comment by nadavkav** [comment]
Yes. I saw the JS code. but I think it should be a cron php function, since I do not trust the JS which might not get executed if the page had some other falty JS that prevents the browser from getting to the activity's JS. (Unfortunately, I get those issue from time to time)

**Comment by garemoko** [comment]
@nadavkav thanks - that code looks like it should resolve the issue @davidpesce reported. @davidpesce - are you able to retest to make sure completion is only triggering when it should? 

On the JS timer, agreed it's not an ideal solution; it also leads to a lot of calls to the LRS. The cron solution might also be less calls as you could get a bunch of statements about ALL users in one hit and then cycle through that. 

Still, as I mentioned, an even better solution is to have an endpoint that can receive incoming statements from an LRS and mark completion of activities. Ideally this would be a separate plugin that adds fields to the global completion settings so that ALL activities have the facility to be marked complete via incoming Tin Can statements.

**Comment by davidpesce** [comment]
Unfortunately, I only see "Course module viewed" and "Activity launched".

This is with both the TC Golf Example and a SL2 file.

**Comment by nadavkav** [comment]
@davidpesce , the SL2 story_html5.html does not work (send statements) but the flash version (story.html) does work. Can you confirm?

**Comment by nadavkav** [comment]
@garemoko, while trying to figure out the best way for the LMS to be informed about the user's activity progress, I just read your posts and other post too, on the LMS/LRS connectivity discussion https://groups.google.com/a/adlnet.gov/forum/#!topic/xapi-design/K2JK54tzer0 . and I must say I am a little amazed that no one from the group of people that put together the xAPI spec thought about the connectivity between the LMS and the LRS thoroughly. Reading the CMI5 draft spec seems to me a better solution we should follow. What do you think?

**Comment by davidpesce** [comment]
@nadavkav Yes, you are correct. The story_html5.html does not send statements. I actually get an "unable to connect to server" error when I launch it.

**Comment by garemoko** [comment]
I believe that in terms of logging, "Course module viewed" and "Activity launched" are what we expect. @nadavkav @scara - does that sound right?

In which case I think this is good for me to do a final check myself and merge?

I am slightly worried about the storyline package though. Can somebody try with the [golf prototype](http://tincanapi.com/prototypes-getting-started/)?

**Comment by davidpesce** [comment]
I tried with both the SL and Golf one. The TC Golf works exactly the same.

**Comment by davidpesce** [comment]
I should clarify.
SL2 - Flash version - works as expected.
SL2 - HTML5 version - doesn't work at all (separate issue)
TC Golf prototype - works as expected.

**Comment by nadavkav** [comment]
Same here. 
SL2 - Flash version - works as expected.
SL2 - HTML5 version - doesn't work at all.
TC Golf prototype - works as expected. (*)

(*) Except, TC Golf does not store the quiz results (last page on the activity) on the WAX LRS that I am using for testing. So I am not sure if this is lack of permissions on my behalf with that specific user or lack of support for Activity API on the LRS.

What LRSes are you using?

**Comment by davidpesce** [comment]
I'm using LearningLocker v1.3.4. 

I hadn't looked that far at the golf example. But you're right, it doesn't show pass/failed or a score. I can see each question answered (not whether it's correct or not). Then a completed statement for the assessment and then a completed statement for the course.

**Comment by garemoko** [comment]
Just a quick apology that I've not merged this yet. I want to test this myself before I merge it, but just to be clear - as far as I'm aware that's just a formality and this is good to merge. It's just waiting for me to get the time to do that review and press the green button.

**Comment by davidpesce** [comment]
Let me know. I've been working on this branch to address issues. If all goes well, I'll have 3 of them knocked off the list.

**Comment by nadavkav** [comment]
No need for apologies :-) Take your time. (I was busy too) just do not forget to squash all the commits into a nice single commit before you merge it.

**Comment by garemoko** [comment]
Just did a test of this and I'm getting a log entry that the user completed the activity when in fact they only launched it. 

Looks like the problem is in complete.php, which should follow the same structure as completion_check.php (remember you had to move the completion notification within the if block). 

It may be worth refactoring that code to avoid duplication so the code only needs to be changed in one place in future.

**Comment by garemoko** [comment]
On squashing the commits - I'd rather retain the history.

**Comment by nadavkav** [comment]
complete.php is called only when the user returns to the LMS (Moodle) and click the "return to course" link. So it is a little weird that you get the completed event when you only launched the activity.

I have moved the event triggering into the IF block. but I can not (yet) upload a fix since I am working in parallel on grades integration into the gradebook and also on the get activity ID from the tincan.xml features.

I will clean and separate the features in my code and upload all the proper commits soon. so let's not close/integrate this one yet.

**Comment by davidpesce** [comment]
@nadavkav Are you sticking with version 03022015 for all your recent changes? I need to change the version number for my changes and didn't want to get mixed up.

**Comment by garemoko** [comment]
Thinking about it - leave the version number as that's likely to cause a merge conflict. I'll update the version number and do a release to the plugins directory once these two PRs are merged. 

Apologies for any confusion caused!

**Comment by davidpesce** [comment]
We still may need it for the upgrade.php file. Or we can simply leave it as xxxxxx as the version number (basically what comes out of xmldb editor)

**Comment by nadavkav** [comment]
Yes. I did not need to change it any more for my other features.

**Comment by davidpesce** [comment]
OK, cool. I'm putting a stake in the sand for Monday (2015032300) as my version.

I found a small typo in the current upgrade.php file. The second if statement has the wrong version number. I fixed it in my last commit (should be in the PR).

**Comment by nadavkav** [comment]
I think I fixed the last issue (moved the "activity completed" event trigger into the completion IF block)

**Comment by nadavkav** [comment]
A little bit off topic but still related...
Any chance we can develop an LRS inside this plugin and have an observer function for the completion event? (or integrate your other LRS project https://github.com/garemoko/PHPCan into this one?)

**Comment by nadavkav** [comment]
Just had an idea...
How about we relay (proxy) the LRS statements to a real LRS, and while launching the activity we will set its LRS url to our LMS (Moodle), we will "listen" to the statements and forward them to the real LRS, and observe the ones we need, like the activity completion and maybe some other useful ones.
This way, we do not need to implement a full LRS. just a "relay" LRS. 
What do you think?

**Comment by davidpesce** [comment]
I think one of the features of an LRS is the ability to forward statements. So this may be possible on the LRS end as well.

**Comment by nadavkav** [comment]
I read on various LRS vendors feature list that this is only implemented on a few and that it seems like a feature outside of the spec. ( Do you have a link to the LRS specs?)

**Comment by garemoko** [comment]
@nadavkav I don't think we need to forward the statements on, just listen for them.
1. Implement the statements POST endpoint
2. receive statements forwarded FROM an LRS
3. compare statement to completion rules to trigger completion
   4, discard statement

If you edit completion settings and need to look at past events, you GET all statements from the LRS as  a one-off move.

**Comment by garemoko** [comment]
Looks like this should work now, testing.

**Comment by nadavkav** [comment]
@garemoko Your idea seems great. though I am not sure we could trust any LRS would forward statements to us (LMS). Only a few support it now, and also if they did, it would probably require the LRS admin to setup the forwarding, Which might complicate things for a simple teacher that just uploaded an activity link to the LMS.

I was actually thinking that we should launch the activity with "our" LRS which will be forwarding statements to the real LRS.

**Pull request merged**